Add dumpers for new loop data structures

BruceForstall opened the pull request:
Specifically, add: `FlowGraphNaturalLoops::Dump()` and `FlowGraphNaturalLoop::Dump()`.

Add corresponding debugger helpers `dNewLoops()` (dumps `Compiler::m_loops`), `dNewLoopsA()` (takes a `FlowGraphNaturalLoops*` argument), `dNewLoop()` (takes a `FlowGraphNaturalLoop*` argument), and their corresponding `c` versions.

A simple dump looks like:
```
***************  (New) Natural loop graph
L00 header: BB05
  Members (3): [BB05..BB07]
  Entry: BB04 -> BB05
  Exit: BB07 -> BB08
  Back: BB07 -> BB05
```

The loop block members are displayed as a block range if the loop blocks happen to all be lexically contiguous without any non-loop blocks in the range.

If there are one or more non-loop blocks in the `bbNext` order range, then a list of ranges is displayed, e.g.:
```
L00 header: BB02
  Members (39): [BB02..BB12];[BB14..BB17];[BB19..BB21];[BB23..BB24];[BB26..BB28];[BB30..BB32];BB34;[BB62..BB72];BB74
  Entry: BB01 -> BB02
  Exit: BB23 -> BB57; BB24 -> BB25; BB20 -> BB36; BB21 -> BB22; BB15 -> BB58; BB16 -> BB59; BB17 -> BB18; BB11 -> BB47; BB12 -> BB13; BB09 -> BB35; BB10 -> BB35; BB10 -> BB61; BB10 -> BB60; BB10 -> BB48; BB10 -> BB56; BB32 -> BB33; BB28 -> BB29; BB74 -> BB75; BB72 -> BB73
  Back: BB28 -> BB02; BB32 -> BB02; BB34 -> BB02; BB69 -> BB02; BB70 -> BB02
```

Otherwise, the full set of loop blocks is displayed individually.

A more complicated example:
```
***************  (New) Natural loop graph
L00 header: BB02
  Members (118): [BB02..BB20];[BB22..BB24];[BB26..BB29];[BB31..BB35];[BB37..BB123]
  Entry: BB01 -> BB02
  Exit: BB123 -> BB124; BB29 -> BB30; BB31 -> BB128; BB45 -> BB130; BB42 -> BB130; BB38 -> BB130; BB35 -> BB36; BB24 -> BB25; BB20 -> BB21
  Back: BB123 -> BB02
L01 header: BB03 parent: L00
  Members (116): [BB03..BB20];[BB22..BB24];[BB26..BB29];[BB31..BB35];[BB37..BB122]
  Entry: BB02 -> BB03
  Exit: BB14 -> BB123; BB15 -> BB123; BB29 -> BB30; BB31 -> BB128; BB45 -> BB130; BB42 -> BB130; BB38 -> BB130; BB35 -> BB36; BB24 -> BB25; BB20 -> BB21
  Back: BB121 -> BB03
L02 header: BB04 parent: L01
  Members (103): [BB04..BB12];[BB14..BB20];[BB22..BB24];[BB26..BB29];[BB31..BB35];[BB37..BB111]
  Entry: BB03 -> BB04
  Exit: BB14 -> BB123; BB15 -> BB123; BB29 -> BB30; BB31 -> BB128; BB45 -> BB130; BB42 -> BB130; BB38 -> BB130; BB35 -> BB36; BB24 -> BB25; BB20 -> BB21; BB11 -> BB122; BB12 -> BB13
  Back: BB111 -> BB04
L03 header: BB05 parent: L02
  Members (17): [BB05..BB12];[BB14..BB16];[BB18..BB20];[BB22..BB24]
  Entry: BB04 -> BB05
  Exit: BB14 -> BB123; BB15 -> BB123; BB16 -> BB17; BB24 -> BB25; BB20 -> BB21; BB11 -> BB122; BB12 -> BB13
  Back: BB24 -> BB05
L04 header: BB06 parent: L03
  Members (13): [BB06..BB12];[BB14..BB16];[BB18..BB20]
  Entry: BB05 -> BB06
  Exit: BB14 -> BB123; BB15 -> BB123; BB16 -> BB17; BB19 -> BB22; BB20 -> BB21; BB11 -> BB122; BB12 -> BB13
  Back: BB20 -> BB06
L05 header: BB07 parent: L04
  Members (10): [BB07..BB12];[BB14..BB16];BB18
  Entry: BB06 -> BB07
  Exit: BB14 -> BB123; BB15 -> BB123; BB16 -> BB17; BB18 -> BB19; BB11 -> BB122; BB12 -> BB13
  Back: BB18 -> BB07
L06 header: BB08 parent: L05
  Members (8): [BB08..BB12];[BB14..BB16]
  Entry: BB07 -> BB08
  Exit: BB14 -> BB123; BB15 -> BB123; BB16 -> BB17; BB10 -> BB18; BB11 -> BB122; BB12 -> BB13
  Back: BB16 -> BB08
L07 header: BB09 parent: L06
  Members (4): [BB09..BB12]
  Entry: BB08 -> BB09
  Exit: BB09 -> BB14; BB10 -> BB18; BB11 -> BB122; BB12 -> BB13
  Back: BB12 -> BB09
L08 header: BB58 parent: L02
  Members (34): [BB58..BB91]
  Entry: BB57 -> BB58
  Exit: BB91 -> BB92
  Back: BB91 -> BB58
```

BruceForstall committed:
Add dumpers for new loop data structures

Specifically, add: `FlowGraphNaturalLoops::Dump()` and
`FlowGraphNaturalLoop::Dump()`.

Add corresponding debugger helpers `dNewLoops()` (dumps
`Compiler::m_loops`), `dNewLoopsA()` (takes a `FlowGraphNaturalLoops*` argument),
`dNewLoop()` (takes a `FlowGraphNaturalLoop*` argument), and their corresponding
`c` versions.

A simple dump looks like:
```
***************  (New) Natural loop graph
L00 header: BB05
  Members (3): [BB05..BB07]
  Entry: BB04 -> BB05
  Exit: BB07 -> BB08
  Back: BB07 -> BB05
```

The loop block members are displayed as a block range if the loop blocks happen
to all be lexically contiguous without any non-loop blocks in the range. Otherwise,
the full set of loop blocks is displayed.

A more complicated example:
```
***************  (New) Natural loop graph
L00 header: BB03
  Members (31): BB03;BB04;BB05;BB06;BB07;BB08;BB09;BB10;BB11;BB12;BB13;BB14;BB15;BB16;BB17;BB18;BB19;BB21;BB22;BB23;BB24;BB25;BB26;BB27;BB28;BB29;BB30;BB32;BB33;BB35;BB36
  Entry: BB01 -> BB03; BB02 -> BB03
  Exit: BB05 -> BB41; BB06 -> BB40; BB07 -> BB39; BB08 -> BB38; BB12 -> BB41; BB14 -> BB37; BB18 -> BB20; BB25 -> BB31; BB32 -> BB34; BB35 -> BB37
  Back: BB36 -> BB03
L01 header: BB05 parent: L00
  Members (5): [BB05..BB09]
  Entry: BB04 -> BB05
  Exit: BB05 -> BB41; BB06 -> BB40; BB07 -> BB39; BB08 -> BB38; BB09 -> BB10
  Back: BB09 -> BB05
L02 header: BB12 parent: L00
  Members (2): [BB12..BB13]
  Entry: BB11 -> BB12
  Exit: BB12 -> BB41; BB13 -> BB14
  Back: BB13 -> BB12
L03 header: BB23 parent: L00
  Members (5): [BB23..BB27]
  Entry: BB22 -> BB23
  Exit: BB25 -> BB31; BB27 -> BB28
  Back: BB27 -> BB23
```

BruceForstall commented:
@AndyAyersMS @jakobbotsch PTAL
cc @dotnet/jit-contrib

jakobbotsch commented on src/coreclr/jit/flowgraph.cpp:
```diff
@@ -4619,6 +4621,181 @@ bool FlowGraphNaturalLoops::FindNaturalLoopBlocks(FlowGraphNaturalLoop* loop, Ar
+    //
+    // Lexicality depends on properly renumbered blocks, which we might not have when dumping.
+
+    printf("\n  Members (%u): ", loop->NumLoopBlocks());
+
+    const unsigned numBlocks = loop->NumLoopBlocks();
```
```suggestion
    const unsigned numBlocks = loop->NumLoopBlocks();
    
    printf("\n  Members (%u): ", numBlocks);

```

jakobbotsch commented on src/coreclr/jit/flowgraph.cpp:
```diff
@@ -4619,6 +4621,181 @@ bool FlowGraphNaturalLoops::FindNaturalLoopBlocks(FlowGraphNaturalLoop* loop, Ar
+        // Add all the loop blocks to a set, then check all the blocks in the identified lexical range.
+        // If there are non-loop blocks found, or if we don't find all the loop blocks in the lexical walk
+        // (meaning the bbNums might not be properly ordered), we fail.
+        loop->VisitLoopBlocks([&](BasicBlock* loopBlock) {
+            BitVecOps::AddElemD(&loopTraits, outputBlocks, loopBlock->bbNewPostorderNum);
+            return BasicBlockVisit::Continue;
+        });
+        bool lexicallyDense = true; // assume the best
+        for (BasicBlock* block = lexicalTopBlock; (block != nullptr) && (block != lexicalEndIteration);
+             block             = block->Next())
+        {
+            if (!loop->ContainsBlock(block))
+            {
+                lexicallyDense = false;
+                break;
+            }
+            BitVecOps::RemoveElemD(&loopTraits, outputBlocks, block->bbNewPostorderNum);
+        }
+        if (!BitVecOps::IsEmpty(&loopTraits, outputBlocks))
+        {
+            lexicallyDense = false;
+        }
```
Instead of the bit vector I think we can use the count of loop blocks seen and compare it to `numBlocks`.

AndyAyersMS replied:
How hard would it be to use range notation for each consecutive span of set bits?

AndyAyersMS replied:
I would just dump things in `bbNum` order as the new loop structures are `bbNum` dependent. If we want something semi-durable (eg is this the "same" loop we saw earlier?) we should perhaps be thinking of dumping the `bbID`.

BruceForstall replied:
> the new loop structures are bbNum dependent

I don't think that's true: they use a new `bbNewPostorderNum` as the key to the loop member bitset, so `bbNum` ordering is only used for "legacy" lexicality uses.

AndyAyersMS replied:
Ah right, that's something Jakob changed from the old synthesis only version....  we can just go with what you have now and see how well it holds up.

BruceForstall replied:
I went ahead and implemented case (1), above. It works out pretty well most of the time.

When we build loops for the alignment placing phase, the `bbNum` are very unordered w.r.t. `bbNext`, so we end up falling back to just displaying all the loop block individually.

jakobbotsch commented on src/coreclr/jit/flowgraph.cpp:
```diff
@@ -4619,6 +4621,181 @@ bool FlowGraphNaturalLoops::FindNaturalLoopBlocks(FlowGraphNaturalLoop* loop, Ar
+        BasicBlock* const exitBlock = edge->getSourceBlock();
+        printf("%s" FMT_BB " ->", first ? "" : "; ", exitBlock->bbNum);
+        exitBlock->VisitRegularSuccs(comp, [=](BasicBlock* succ) {
+            if (!loop->ContainsBlock(succ))
```
```suggestion
            if (fgGetPredForBlock(succ, exitBlock) == edge)
```
Otherwise the output will look strange for a `BBJ_SWITCH` with multiple exits, I think.

We should probably store the target block as part of the exit edges (until `FlowEdge` begins storing them).

A small terminology nit -- the LLVM/GCC terminology of loops call the `exitBlock` here for "exiting" blocks, while the actual "exit" block is the one outside the loop (`succ` in this case).

jakobbotsch reviewed (commented):
This looks great, thanks for working on it.

jakobbotsch commented on src/coreclr/jit/flowgraph.cpp:
```diff
@@ -4619,6 +4621,181 @@ bool FlowGraphNaturalLoops::FindNaturalLoopBlocks(FlowGraphNaturalLoop* loop, Ar
+    printf("\n  Members (%u): ", loop->NumLoopBlocks());
+
+    const unsigned numBlocks = loop->NumLoopBlocks();
+    if (numBlocks == 0)
+    {
+        // This should never happen?
+        printf("NONE?");
+    }
```
Indeed, this can be an assert.

AndyAyersMS replied:
Asserts in diagnostic helpers are annoying, so I'd keep this as is.

BruceForstall replied:
Yes, dumpers should not have asserts, but should be written to be (reasonably) tolerant of buggy data structures (presumably there would be an equivalent assert someplace else).

jakobbotsch commented on src/coreclr/jit/flowgraph.cpp:
```diff
@@ -4619,6 +4621,181 @@ bool FlowGraphNaturalLoops::FindNaturalLoopBlocks(FlowGraphNaturalLoop* loop, Ar
+    }
+    else if (numBlocks == 1)
+    {
+        // Should be able to assume the only block is the header, but just in case, use the normal visitor.
+        loop->VisitLoopBlocksReversePostOrder([](BasicBlock* block) {
+            printf(FMT_BB " ", block->bbNum);
+            return BasicBlockVisit::Continue;
+        });
```
Also correct that this will always be just the header.

BruceForstall committed:
Implement list of ranges for block member dump

List of ranges compacts the dump when there is just one or a small number
of non-loop blocks in the `bbNext` order range.

Also, address some code review feedback.

BruceForstall commented:
@AndyAyersMS @jakobbotsch This is ready for another look

jakobbotsch reviewed (approved):
Looks great to me.
